### PR TITLE
Refactored C++ code generator to not directly depend on protobuf

### PR DESCRIPTION
**Pull request by gwvo**
This opens the door for other serializers (such as e.g. FlatBuffers)
to share this code generator when using GRPC.

cpp_generator.h/.cc are now stand-alone, by abstracting all protobuf specifics
into interfaces which are implemented for protobuf in cpp_plugin.cc

### CODE REVIEW

**Comment by ctiller** [comment]
This is OK to test

**Comment by gwvo** [comment]
Ok, fixed one silly unique_ptr bug :)

**Comment by gwvo** [comment]
Some tests failed because some configurations use `-Wunused-private-field` and others do not (certainly the local build did not). Might we worth fixing.

**Comment by gwvo** [comment]
Now one test failed on a `Timeout`. Is this just the build machine being flakey? I certainly hope my PR is not breaking Python tests :)

    Running       beta_python_plugin_test.PythonPluginTest.testFullDuplexCall
    [95mTesting gRPC Python...
    [0m[92mSUCCESS       beta_python_plugin_test.PythonPluginTest.testFullDuplexCall
    [0mRunning       beta_python_plugin_test.PythonPluginTest.testFullDuplexCallCancelled

**Comment by ctiller** [comment]
It's the Python tests that are wrong.

File: src/compiler/cpp_generator.h [hunk 1]
```diff
@@ -48,37 +64,80 @@ struct Parameters {
   grpc::string grpc_search_path;
 };
 
+// An abstract interface representing a method.
+struct Method {
+  virtual ~Method() {};
```
Comment (yang-g) [inline src/compiler/cpp_generator.h]: Remove extra `;`. There are a couple below as well.

**Comment by yang-g** [comment]
Thanks for the contribution. 
LGTM other than nits.

**Comment by gwvo** [comment]
Fixed the nits...

**Comment by yang-g** [comment]
LGTM
waiting for green tests

**Comment by nicolasnoble** [comment]
```
src/compiler/cpp_plugin.cc:110:17: error: private field 'str_' is not used [-Werror,-Wunused-private-field]
  grpc::string *str_;
```

**Comment by nicolasnoble** [comment]
Nevermind me, my web ui hasn't been refreshing properly ?

**Pull request merged**